File: examples/julia/FromSamples2D_banana.jl
using MParT, CxxWrap
using Distributions, LinearAlgebra, Statistics, Optimization, OptimizationOptimJL

make_plot = true

if make_plot
    using GLMakie
end
##
num_points = 1000
z = randn(2,num_points)
x1 = z[1,:]
x2 = z[2,:] + z[1,:].^2
x = collect([x1 x2]')

test_n_pts = 10_000
test_z = randn(2,test_n_pts)
test_x1 = test_z[1,:]
test_x2 = test_z[2,:] + test_z[1,:].^2
test_x = collect([test_x1 test_x2]')

# For Plotting and computing reference density
reference_density = MvNormal(I(2))
t = range(-5,5,length=100)
reference_density_pdf = [pdf(reference_density, [t1,t2]) for t1 in t, t2 in t]

## Set up map and initialize coefficients
opts = MapOptions()
tri_map = CreateTriangular(2,2,2,opts)
coeffs = zeros(numCoeffs(tri_map))

function obj(coeffs, p)
    tri_map, x = p
    SetCoeffs(tri_map, coeffs)
    map_of_x = Evaluate(tri_map, x)
    ref_density_of_map_of_x = logpdf(reference_density, map_of_x)
    log_det = LogDeterminant(tri_map, x)
    -sum(ref_density_of_map_of_x + log_det)/num_points
end

function grad_obj(g, coeffs, p)
    tri_map, x = p
    SetCoeffs(tri_map, coeffs)
    map_of_x = Evaluate(tri_map, x)
    grad_ref_density_of_map_of_x = -CoeffGrad(tri_map, x, map_of_x)
    grad_log_det = LogDeterminantCoeffGrad(tri_map, x)
    g .= -sum(grad_ref_density_of_map_of_x + grad_log_det, dims=2)/num_points
end

## Plot before Optimization
map_of_x = Evaluate(tri_map, x)
if make_plot
    fig = Figure()
    ax1 = Axis(fig[1,1], title="Before Optimization")
    contour!(ax1, t, t, reference_density_pdf)
    scatter!(ax1, test_x[1,:], test_x[2,:], color=(:blue,0.5), label="Target Samples")
    axislegend(ax1)
    display(fig)
end

u0 = CoeffMap(tri_map)
p = (tri_map, x)
fcn = OptimizationFunction(obj, grad = grad_obj)
prob = OptimizationProblem(fcn, u0, p, g_tol = 1e-16)

## Optimize


println("Starting coeffs")
println(u0)
println("and error: $(obj(u0,p))")
println("===================")
sol = solve(prob, BFGS())
##
u_final = sol.u
SetCoeffs(tri_map, u_final)
println("Final coeffs")
println(u_final)
println("and error: $(obj(u_final,p))")
println("===================")
map_of_test_x = Evaluate(tri_map, test_x)
if make_plot
    fig = Figure()
    ax2 = Axis(fig[1,1], title="After Optimization")
    contour!(ax2, t, t, reference_density_pdf)
    scatter!(ax2, map_of_test_x[1,:], map_of_test_x[2,:], color=(:blue,0.5), label="Target Samples")
    axislegend(ax2)
    display(fig)
end
mean_of_map = mean(map_of_test_x, dims=2)
cov_of_map = cov(map_of_test_x, dims=2)
println("Mean of map")
println(mean_of_map)
println("Covariance of map")
println(cov_of_map)

File: examples/python/FromSamples2D_banana.py
# # Construct map from banana distribution samples in 2D


from mpart import *
import numpy as np
from scipy.optimize import minimize
from scipy.stats import multivariate_normal
import matplotlib.pyplot as plt


# Make target samples for training
num_points = 1000
z = np.random.randn(2,num_points)
x1 = z[0]
x2 = z[1] + z[0]**2
x = np.vstack([x1,x2])


# Make target samples for testing
test_z = np.random.randn(2,10000)
test_x1 = test_z[0]
test_x2 = test_z[1] + test_z[0]**2
test_x = np.vstack([test_x1,test_x2])


# For plotting and computing reference density 
ref_distribution = multivariate_normal(np.zeros(2),np.eye(2))  #standard normal
t = np.linspace(-5,5,100)
grid = np.meshgrid(t,t)
ref_pdf_at_grid = ref_distribution.pdf(np.dstack(grid))


# Set-up map and initize map coefficients
map_options = MapOptions()
transport_map = CreateTriangular(2,2,2,map_options)
coeffs = np.zeros(transport_map.numCoeffs)
transport_map.SetCoeffs(coeffs)


# KL divergence objective
def obj(coeffs, transport_map, x):
    transport_map.SetCoeffs(coeffs)
    map_of_x = transport_map.Evaluate(x)
    ref_logpdf = ref_distribution.logpdf(map_of_x.T)
    log_det = transport_map.LogDeterminant(x)
    return -np.sum(ref_logpdf + log_det)/num_points


def grad_obj(coeffs, transport_map, x):
    transport_map.SetCoeffs(coeffs)
    map_of_x = transport_map.Evaluate(x)
    grad_ref_logpdf = -transport_map.CoeffGrad(x, map_of_x)
    grad_log_det = transport_map.LogDeterminantCoeffGrad(x)
    return -np.sum(grad_ref_logpdf + grad_log_det, 1)/num_points


# Before optimization plot
plt.figure()
plt.contour(*grid, ref_pdf_at_grid)
plt.scatter(test_x[0],test_x[1], facecolor='blue', alpha=0.1, label='Target samples')
plt.legend()
plt.show()


# Print initial coeffs and objective
print('Starting coeffs')
print(transport_map.CoeffMap())
print('and error: {:.2E}'.format(obj(transport_map.CoeffMap(), transport_map, x)))
print('==================')


# Optimize
optimizer_options={'gtol': 1e-4, 'disp': True}
res = minimize(obj, transport_map.CoeffMap(), args=(transport_map, x), jac=grad_obj, method='BFGS', options=optimizer_options)


# Print final coeffs and objective
print('Final coeffs')
print(transport_map.CoeffMap())
print('and error: {:.2E}'.format(obj(transport_map.CoeffMap(), transport_map, x)))
print('==================')


# After optimization plot
map_of_test_x = transport_map.Evaluate(test_x)
plt.figure()
plt.contour(*grid, ref_pdf_at_grid)
plt.scatter(map_of_test_x[0],map_of_test_x[1], facecolor='blue', alpha=0.1, label='Normalized samples')
plt.legend()
plt.show()


# Print statistics of normalized samples (TODO replace with better Gaussianity check)
print('==================')
mean_of_map = np.mean(map_of_test_x,1)
print("Mean of normalized test samples")
print(mean_of_map)
print('==================')
print("Cov of normalized test samples")
cov_of_map = np.cov(map_of_test_x)
print(cov_of_map)